File: packages/sdk-ts/src/lineage.ts
/** Object-identity registry for cross-hop data-flow lineage (source_hop_ids). */

const PREVIEW_MAX = 500;

export function truncatePreview(text: string, max = PREVIEW_MAX): string {
  const line = text.replace(/\s+/g, ' ').trim();
  return line.length <= max ? line : `${line.slice(0, max)}…`;
}

export function previewFromValue(value: unknown): string | undefined {
  if (value === null || value === undefined) return undefined;
  if (typeof value === 'string') {
    const t = value.trim();
    return t ? truncatePreview(t) : undefined;
  }
  if (typeof value === 'object') {
    try {
      const s = JSON.stringify(value);
      return s && s !== '{}' ? truncatePreview(s) : undefined;
    } catch {
      return undefined;
    }
  }
  return truncatePreview(String(value));
}

/** Registry keyed by object identity with string fallback (mirrors VerdictLens source_span_ids). */
export class HopLineageRegistry {
  private readonly objects = new WeakMap<object, string>();
  private readonly strings = new Map<string, string>();
  /** Strong refs so short-lived values are not GC'd before the run ends. */
  private readonly refs: unknown[] = [];

  register(value: unknown, hopId: string): void {
    if (value === null || value === undefined) return;
    if (typeof value === 'object') {
      this.objects.set(value, hopId);
      this.refs.push(value);
    } else if (typeof value === 'string' && value.trim()) {
      this.strings.set(value, hopId);
      this.refs.push(value);
    }
  }

  detectSources(args: unknown[]): string[] {
    const seen = new Set<string>();
    const out: string[] = [];
    for (const arg of args) {
      if (arg !== null && typeof arg === 'object') {
        const id = this.objects.get(arg);
        if (id && !seen.has(id)) {
          seen.add(id);
          out.push(id);
        }
      }
      if (typeof arg === 'string') {
        const id = this.strings.get(arg);
        if (id && !seen.has(id)) {
          seen.add(id);
          out.push(id);
        }
      }
    }
    return out;
  }
}

File: packages/sdk-py/src/blamr_sdk/lineage.py
"""Object-identity registry for source_hop_ids lineage."""

from __future__ import annotations

import json
from typing import Any


def preview_from_value(value: Any) -> str | None:
  if value is None:
    return None
  if isinstance(value, str):
    t = value.strip()
    return t if t else None
  if isinstance(value, (dict, list)):
    try:
      s = json.dumps(value, default=str)
      return s if s and s != "{}" else None
    except Exception:
      return None
  return str(value)


class HopLineageRegistry:
  def __init__(self) -> None:
    self._objects: dict[int, str] = {}
    self._strings: dict[str, str] = {}
    self._refs: list[Any] = []

  def register(self, value: Any, hop_id: str) -> None:
    if value is None:
      return
    if isinstance(value, str) and value.strip():
      self._strings[value] = hop_id
      self._refs.append(value)
    elif isinstance(value, (dict, list, tuple)):
      self._objects[id(value)] = hop_id
      self._refs.append(value)

  def detect_sources(self, args: tuple[Any, ...], kwargs: dict[str, Any] | None = None) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    values: list[Any] = list(args)
    if kwargs:
      values.extend(kwargs.values())
    for val in values:
      if isinstance(val, str):
        hop_id = self._strings.get(val)
        if hop_id and hop_id not in seen:
          seen.add(hop_id)
          out.append(hop_id)
      elif isinstance(val, (dict, list, tuple)):
        hop_id = self._objects.get(id(val))
        if hop_id and hop_id not in seen:
          seen.add(hop_id)
          out.append(hop_id)
    return out
